### ISSUE TITLE
separate go raml parser to its own project

### ISSUE THREAD

**Issue by andjarnic** [Jumpscale/go-raml#109]
As the subject implies, it would be good if the Go parser for RAML 1.0 were its own project, and this project then depended upon it.. or better yet, given the name of this project, having this project only be a GO parser and build other Github projects on top of it, such as Go-Raml-Server, Go-Raml-Client, etc.

**Comment by robvanmieghem** [comment]
@andjarnic agreed, even though it is considered lower priority for us at the moment

**Comment by zaibon** [comment]
@iwanbk how hard would it be to extract the parser with the current status of the code base ?
I think that would be a good step in the direction for template support too.

**Comment by iwanbk** [comment]
> @iwanbk how hard would it be to extract the parser with the current status of the code base ?

If only extracting, then the parser is on separate package, so quite easy to extract.

However this parser is not a mere parser, it added few things to make the code generator easier to generate the code. So, we might need to extract this part of code, if we want to make it as mere parser.

An example of added things is when the parses see `inline type declaration`. The parser will generates new `type` based on that inline declaration, and then put that type into `Types` field of the RAML object.

It is of course doable to separate this kind of code, but i don't see the point to separate it:
- there is already raml parser out there (https://github.com/go-raml/raml) which our parser based on
- why suddenly we want to separate it while before this we pull the https://github.com/go-raml/raml parser into this project

Except that we want to make it parser for code generator, i can't see the point.

> I think that would be a good step in the direction for template support too.

Do you mean this? https://github.com/raml-org/raml-spec/blob/master/versions/raml-10/raml-10.md#template-uris-and-uri-parameters

How it relates?

**Comment by kevinduffey** [comment]
my original ask was to see it separated so that go programmers could use it to generate whatever they wanted.. be it a new project they had in mind for a raml2xxx generator, or otherwise. 

I still hold that separating it so that the parser can be used like a library/sdk in some fashion, be it a stream processing parser or hooks or something.. it would be ideal for this to be a stand alone parser library, that can be used by other projects (including this generator).

**Comment by zaibon** [comment]
https://github.com/go-raml/raml seems to only supports raml 0.8 and doesn't look maintained anymore. So I think it's still make sense to promote our implementation of the parser. Since it's supports 1.0 and we're still actively using it.

I agree with @kevinduffey if we can have a generic parser on which we can build on top. It gives us flexibility to what's we create from it.

Regarding template, I talking about providing our templates for sever and client generation. So we can tune the generated code. This is also something we'll probably need in a near future for our project.

**Comment by kevinduffey** [comment]
To  add to this.. Raml 1 annotations allow custom coding possibilities that are otherwise difficult to realize without the ability to be part of the parsing process. RAML2JAXRS adds a custom extension path to allow java developers to add custom annotations and the code that processes them, which is nice.. though a little complicated. Adding something like that.. (forgive me as I am still not able to dive in on Go and learn it yet so not sure if this exists already) or simply breaking the parser out allowing developers to do as they will with those sort of things is a big plus.

**Comment by iwanbk** [comment]
> https://github.com/go-raml/raml seems to only supports raml 0.8 and doesn't look maintained anymore. So I think it's still make sense to promote our implementation of the parser. Since it's supports 1.0 and we're still actively using it.

Yeah, make sense then

> Regarding template, I talking about providing our templates for sever and client generation. So we can tune the generated code. This is also something we'll probably need in a near future for our project.

Not sure what you are talking about, I'm sure you already know that we already use template for code generation.

**Comment by zaibon** [comment]
> Not sure what you are talking about, I'm sure you already know that we already use template for code generation.

There is no way for the user to provide its own template. Now the template are hard-coded on the code of go-raml.

**Comment by kevinduffey** [comment]
So.. if you all dare to do a little digging at a javascript project, take a look at raml-generator. It is written in node I believe.. but it basically parses the RAML 0.8 and applies it to templates you can provide yourself to generate whatever you want. It allows the developer of the tool (parser) to focus more on the template generator and less on the semantics of the parsing of raml. You still have to do some parsing though.. and to my point and I believe @zaibon it would be much better to have a dynamic templating capability built in to the parser so that the parser does the bulk of the work.. possibly with some sort of callback or interface or something that a developer would implement to add their own generator output. This could then be used for code generation, documentation, tests, etc.

**Comment by iwanbk** [comment]
@kevinduffey thanks for the pointer

@zaibon could you please set priority and proper milestone for this?
i'd like to work on it after bring multiple status code handling to python and nim client, which include some refactoring to clean some mess.

**Comment by iwanbk** [comment]
> i'd like to work on it after bring multiple status code handling to python and nim client, which include some refactoring to clean some mess.

Also some small issues that still open
